rs/bitcoin/kyt/src/state.rs

```diff
@@ -52,37 +52,107 @@ pub struct FetchedTx {
 // Max number of concurrent http outcalls.
 const MAX_CONCURRENT: u32 = 50;
 
+// Default expiry of cached fetch transaction entries is 7 days (in nano seconds).
+const DEFAULT_FETCH_TX_EXPIRY: u64 = 7 * 24 * 3600 * 1000000;
+
+// Max number of entries in the cache is set to 10_000. Since the average transaction size
+// is about 400 bytes, the estimated memory usage of the cache is in the order of 10s of MBs.
+const MAX_FETCH_TX_ENTRIES: usize = 10_000;
+
 // The internal KYT state includes:
 // 1. Outcall capacity, a semaphore limiting max concurrent outcalls.
 // 2. fetch transaction status, indexed by transaction id.
 //
 // TODO(XC-191): persist canister state
 thread_local! {
     static OUTCALL_CAPACITY: RefCell<u32> = const { RefCell::new(MAX_CONCURRENT) };
-    static FETCH_TX_STATUS: RefCell<BTreeMap<Txid, FetchTxStatus>> = RefCell::new(BTreeMap::default());
+    static FETCH_TX_CACHE: RefCell<FetchTxCache<FetchTxStatus>> = RefCell::new(
+        FetchTxCache::new(DEFAULT_FETCH_TX_EXPIRY, MAX_FETCH_TX_ENTRIES)
+    );
+}
+
+// Time in nanoseconds since the epoch (1970-01-01).
+type Timestamp = u64;
+
+// Time duration in nanoseconds.
+type Duration = u64;
+
+struct FetchTxCache<T> {
+    expiry: Duration,
+    max_entries: usize,
+    status: BTreeMap<Txid, T>,
+    created: VecDeque<(Txid, Timestamp)>,
+}
+
+impl<T> FetchTxCache<T> {
+    fn new(expiry: Duration, max_entries: usize) -> Self {
+        Self {
+            expiry,
+            max_entries,
+            status: BTreeMap::new(),
+            created: VecDeque::new(),
+        }
+    }
+
+    fn get_status(&self, txid: Txid) -> Option<&T> {
+        self.status.get(&txid)
+    }
+
+    // Set the status of a txid with a timestamp. If the txid does not
+    // already have a status, a new status entry is created and associated
+    // with the given the timestamp.
+    //
+    // When called with non-decreasing timestamps, this function will remove
+    // at most one past (i.e. the oldest) entry based on the `expiry` and
+    // `max_entries` settings, when either condition satisfies.
+    fn set_status_with(&mut self, txid: Txid, status: T, now: Timestamp) {
```

Comment:
For testing purposes it might be useful to return a value to indicate which value got ejected and why. E.g.,

```rust
 fn put(&mut self, txid: Txid, status: T, now: Timestamp) -> Vec<EvictedEntry<T>>{}

pub enum EvictedEntry<T> {
  Expired(Txid, T, Timestamp),
  Ejected(Txid, T, Timestamp),
}
```
WDYT?

Reply:
Good idea! Changed the signature to return an Option since we are only purging one entry at a time.